src/Features/Core/Portable/CodeRefactorings/SyncNamespace/AbstractSyncNamespaceService.ChangeNamespaceCodeAction.cs

```diff
@@ -0,0 +1,526 @@
+                {
+                    builder.Add(CreateImport(generator, names[i], withFormatterAnnotation));
+                }
+
+                return builder.ToImmutableAndFree();
+            }
+
+            private SyntaxNode CreateImport(SyntaxGenerator syntaxGenerator, string name, bool withFormatterAnnotation)
+            {
+                var import = syntaxGenerator.NamespaceImportDeclaration(name);
+                if (withFormatterAnnotation)
+                {
+                    import = import.WithAdditionalAnnotations(Formatter.Annotation);
+                }
+                return import;
+            }
+
+            private async Task<(Solution, ImmutableArray<DocumentId>)> ChangeNamespaceToMatchFoldersAsync(Solution solution, DocumentId id, CancellationToken cancellationToken)
+            {
+                var document = solution.GetDocument(id);
+                var targetNamespace = _state.TargetNamespace;
+
+                var declarationRoot = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+                var container = declarationRoot.DescendantNodes().FirstOrDefault(node => node is TNamespaceDeclarationSyntax) ?? declarationRoot;
+
+                var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+                var declaredSymbols = GetDeclaredSymbolsInContainer(semanticModel, container, cancellationToken);
+
+                var documentSet = ImmutableHashSet.Create(document);
+                var editor = await DocumentEditor.CreateAsync(document, cancellationToken).ConfigureAwait(false);
+
+                var refLocationsInCurrentDocument = new List<ReferenceLocation>();
+                var refLocationsInOtherDocuments = new List<ReferenceLocation>();
+
+                foreach (var declaredSymbol in declaredSymbols)
+                {
+                    cancellationToken.ThrowIfCancellationRequested();
+
+                    var refSymbols = await SymbolFinder.FindReferencesAsync(
+                      declaredSymbol,
+                      solution,
+                      cancellationToken).ConfigureAwait(false);
+
+                    var refLocationsForSymbol = refSymbols.Where(refSymbol => refSymbol.Definition == declaredSymbol)
+                        .SelectMany(refSymbol => refSymbol.Locations);
+
+                    foreach (var refLocation in refLocationsForSymbol)
+                    {
+                        if (refLocation.Document.Id == document.Id)
+                        {
+                            refLocationsInCurrentDocument.Add(refLocation);
+                        }
+                        else
+                        {
+                            Debug.Assert(!PathUtilities.PathsEqual(refLocation.Document.FilePath, document.FilePath));
+                            refLocationsInOtherDocuments.Add(refLocation);
+                        }
+                    }
+                }
```

Comment:
consider extracting this all out into a helper.  it will help this method be more clear in terms of what its purpose is and what steps it is taking at a hgh level.